hermes/bin/src/runtime_extensions/hermes/crypto/state.rs

```diff
@@ -0,0 +1,222 @@
+    fn new() -> Self {
+        Self {
+            id_to_resource_map: DashMap::new(),
+            resource_to_id_map: DashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    fn get_next_id(&mut self) -> u32 {
+        let num = self.next_id + 1;
+        self.next_id = num;
+        num
+    }
+
+    //FIXME - This should return a reference?
+    fn get_resource_from_id(&self, id: &u32) -> Option<XPrv> {
+        self.id_to_resource_map.get(id).map(|entry| entry.value().clone())
+    }
+    
+    //FIXME - This should return a reference?
+    fn get_id_from_resource(&self, resource: &XPrv) -> Option<u32> {
+        self.resource_to_id_map
+            .get(&WrappedXPrv::from_xprv(resource.clone()))
+            .map(|entry| entry.value().clone())
+    }
+
+    /// Drop the item from resources using id if possible.
+    /// Return the id of the resource if successful remove from maps
+    fn drop(&mut self, id: u32) -> Option<u32> {
+        if let Some(resource) = self.get_resource_from_id(&id) {
+            if let Some(associated_id) = self.get_id_from_resource(&resource) {
+                if associated_id == id {
+                    if let Some(r) = self.id_to_resource_map.remove(&id) {
+                        self.resource_to_id_map.remove(&WrappedXPrv(r.1));
+                        return Some(associated_id);
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+static CRYPTO_INTERNAL_STATE: Lazy<Arc<State>> = Lazy::new(|| Arc::new(DashMap::new()));
+
+/// Check the state whether the context exists and return the resources if possible.
+fn check_context_and_return_resources(
+    app_name: &String, module_id: &Ulid, event_name: &String, counter: &u64,
+) -> Option<ResourceHolder> {
+    let binding = CRYPTO_INTERNAL_STATE.clone();
+    if let Some(app_map) = binding.get(app_name) {
+        if let Some(module_map) = app_map.get(module_id) {
+            if let Some(event_map) = module_map.get(event_name) {
+                if let Some(counter_map) = event_map.get(counter) {
+                    // FIXME - This shouldn't be clone?
+                    return Some(counter_map.clone());
+                }
+            }
+        }
+    }
+    return None;
+}
+
+#[allow(dead_code)]
+/// Set the state according to the app context.
+pub(crate) fn set_state(app_name: String, module_id: Ulid, event_name: String, counter: u64) {
+    // Counter -> ResourceHolder
+    let counter_map = DashMap::new();
+    counter_map.insert(counter, ResourceHolder::new());
+    // Event -> Counter
+    let event_map = DashMap::new();
+    event_map.insert(event_name, counter_map);
+    // Module -> Event
+    let module_map = DashMap::new();
+    module_map.insert(module_id, event_map);
+    // App -> Module
+    CRYPTO_INTERNAL_STATE.insert(app_name, module_map);
+}
+
+#[allow(dead_code)]
+/// Get the resource from the state using id if possible.
+// FIXME - Should this return a reference?
```

Comment:
Doesn't need to return a reference, for now.  It might be more optimal to return a reference, but don't worry about that now.